src/complete.cpp

```diff
@@ -838,19 +836,24 @@ static bool short_ok(const wcstring &arg, const complete_entry_opt_t *entry,
     return result;
 }
 
+struct cmd_data_t {
+    const wcstring &name;
+    bool has_path;
+};
```

Comment:
Storing a reference inside a struct is always a little frightening, since it may become dangling. If you're up for it, I suggest making this store the string and changing parse_cmd_string() to return a cmd_data_t directly. If that seems too much it's fine as it is, we can refactor it later.